File: Algorithm_LevelOne/Rotate_Array/Rotate_Array.cc
// 189. Rotate Array C++

#include <vector>

class Solution {
public:
    void rotate(std::vector<int>& nums, int k) {
        int length = nums.size();
        k = k % length;
        int start = 0;
        int count = 0;
        while (count < length) {
            int current = start;
            int prev = nums[start];

            while (true) {
                int next = (current + k) % length;
                int tmp = nums[next];
                nums[next] = prev;
                count++;
                prev = tmp;
                current = next;
                if (current == start) break;
            }
            start++;
        }
    }
};

File: Top_Interview_150/Rotate Array.py
# 189. Rotate Array

class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        l = len(nums)
        count = 0
        start = 0
        while count < l:
            current = start
            prev = nums[start]

            while True:
                nxt = (current + k) % l
                nums[nxt], prev = prev, nums[nxt]
                current = nxt
                count += 1
                if start == current: break
            
            start += 1
            